ibis/backends/base/__init__.py

```diff
@@ -369,6 +369,100 @@ def read_csv(
             f"{self.name} does not support direct registration of CSV data."
         )
 
+    @util.experimental
+    def to_parquet(
+        self,
+        expr: ir.Table,
+        path: str | Path,
+        *,
+        params: Mapping[ir.Scalar, Any] | None = None,
+        limit: int | str | None = None,
+        **kwargs: Any,
+    ) -> None:
+        """Write the results of executing the given expression to a parquet file.
+
+        This method is eager and will execute the associated expression
+        immediately.
+
+        Parameters
+        ----------
+        expr
+            The ibis expression to execute and persist to parquet.
+        path
+            The data source. A string or Path to the parquet file.
+        params
+            Mapping of scalar parameter expressions to value.
+        limit
+            An integer to effect a specific row limit. A value of `None` means
+            "no limit". The default is in `ibis/config.py`.
+        **kwargs
+            Additional keyword arguments passed to pyarrow.parquet.ParquetWriter
+
+        https://arrow.apache.org/docs/python/generated/pyarrow.parquet.ParquetWriter.html
+        """
+        self._import_pyarrow()
+        import pyarrow.parquet as pq
+
+        batch_reader = expr.to_pyarrow_batches(params=params, limit=limit)
+
+        writer = pq.ParquetWriter(path, batch_reader.schema)
+
+        try:
+            while batch := batch_reader.read_next_batch():
+                writer.write_batch(batch)
+        except StopIteration:
+            pass
+        finally:
+            writer.close()
+
+    @util.experimental
+    def to_csv(
+        self,
+        expr: ir.Table,
+        path: str | Path,
+        *,
+        params: Mapping[ir.Scalar, Any] | None = None,
+        limit: int | str | None = None,
+        **kwargs: Any,
+    ) -> None:
+        """Write the results of executing the given expression to a CSV file.
+
+        This method is eager and will execute the associated expression
+        immediately.
+
+        Parameters
+        ----------
+        expr
+            The ibis expression to execute and persist to CSV.
+        path
+            The data source. A string or Path to the CSV file.
+        params
+            Mapping of scalar parameter expressions to value.
+        limit
+            An integer to effect a specific row limit. A value of `None` means
+            "no limit". The default is in `ibis/config.py`.
+        **kwargs
+            Additional keyword arguments passed to pyarrow.csv.CSVWriter
+
+        https://arrow.apache.org/docs/python/generated/pyarrow.csv.CSVWriter.html
+        """
+        self._import_pyarrow()
+        import pyarrow.csv as pcsv
+
+        pcsv.write_csv(expr.to_pyarrow(params=params, limit=limit), str(path), **kwargs)
```

Comment:
It looks like we're writing the CSV twice, once using `to_pyarrow` and the other using `to_pyarrow_batches`. Am I reading this code correctly?

Reply:
lol, whoops -- don't know how that reappeared